File: source/services/admin-service/src/routes/chargers.rs
use crate::config;
use crate::db;
use crate::error::AppError;
use crate::models::{ChargerListParams, CreateChargerRequest, UpdateChargerRequest};
use crate::AppState;
use actix_web::{delete, get, post, put, web, HttpRequest, HttpResponse};

const VALID_CONNECTOR_TYPES: &[&str] = &["type2", "type3", "ccs", "chademo"];
const VALID_CHARGER_STATUSES: &[&str] = &["available", "in_use", "maintenance", "offline"];

#[post("/api/chargers")]
pub async fn create_charger(
    req: HttpRequest,
    state: web::Data<AppState>,
    body: web::Json<CreateChargerRequest>,
) -> Result<HttpResponse, AppError> {
    let actor = config::x_partner_id(&req);
    if !VALID_CONNECTOR_TYPES.contains(&body.connector_type.as_str()) {
        return Err(AppError::ValidationError(format!(
            "invalid connector_type '{}', must be one of {:?}",
            body.connector_type, VALID_CONNECTOR_TYPES
        )));
    }
    if body.power_kw <= 0.0 {
        return Err(AppError::ValidationError(
            "power_kw must be greater than 0".to_string(),
        ));
    }
    if let Some(ref status) = body.status {
        if !VALID_CHARGER_STATUSES.contains(&status.as_str()) {
            return Err(AppError::ValidationError(format!(
                "invalid status '{}', must be one of {:?}",
                status, VALID_CHARGER_STATUSES
            )));
        }
    }
    let charger = db::chargers::create_charger(&state.pool, body.into_inner(), &actor).await?;
    Ok(HttpResponse::Created().json(charger))
}

#[get("/api/chargers")]
pub async fn list_chargers(
    state: web::Data<AppState>,
    query: web::Query<ChargerListParams>,
) -> Result<HttpResponse, AppError> {
    let page = query.page.unwrap_or(1).max(1);
    let page_size = query.page_size.unwrap_or(20).min(100).max(1);
    let station_id = query.station_id.as_deref();
    let result = db::chargers::list_chargers(&state.pool, station_id, page, page_size).await?;
    Ok(HttpResponse::Ok().json(result))
}

#[get("/api/chargers/{id}")]
pub async fn get_charger(
    state: web::Data<AppState>,
    path: web::Path<String>,
) -> Result<HttpResponse, AppError> {
    let id = path.into_inner();
    let charger = db::chargers::get_charger(&state.pool, &id).await?;
    Ok(HttpResponse::Ok().json(charger))
}

#[put("/api/chargers/{id}")]
pub async fn update_charger(
    req: HttpRequest,
    state: web::Data<AppState>,
    path: web::Path<String>,
    body: web::Json<UpdateChargerRequest>,
) -> Result<HttpResponse, AppError> {
    let actor = config::x_partner_id(&req);
    let id = path.into_inner();
    if let Some(ref ct) = body.connector_type {
        if !VALID_CONNECTOR_TYPES.contains(&ct.as_str()) {
            return Err(AppError::ValidationError(format!(
                "invalid connector_type '{}', must be one of {:?}",
                ct, VALID_CONNECTOR_TYPES
            )));
        }
    }
    if let Some(pk) = body.power_kw {
        if pk <= 0.0 {
            return Err(AppError::ValidationError(
                "power_kw must be greater than 0".to_string(),
            ));
        }
    }
    if let Some(ref status) = body.status {
        if !VALID_CHARGER_STATUSES.contains(&status.as_str()) {
            return Err(AppError::ValidationError(format!(
                "invalid status '{}', must be one of {:?}",
                status, VALID_CHARGER_STATUSES
            )));
        }
    }
    let charger = db::chargers::update_charger(&state.pool, &id, body.into_inner(), &actor).await?;
    Ok(HttpResponse::Ok().json(charger))
}

#[delete("/api/chargers/{id}")]
pub async fn delete_charger(
    state: web::Data<AppState>,
    path: web::Path<String>,
) -> Result<HttpResponse, AppError> {
    let id = path.into_inner();
    db::chargers::delete_charger(&state.pool, &id).await?;
    Ok(HttpResponse::Ok().json(serde_json::json!({"deleted": true})))
}

File: source/packages/client-core/src/adminApi/chargers.ts
import type {
  AdminChargerDto,
  CreateChargerRequest,
  UpdateChargerRequest,
} from "@bornemap/domain-types";

export interface PaginationInfo {
  page: number;
  per_page: number;
  total: number;
  total_pages: number;
}

export interface PaginatedResponse<T> {
  data: T[];
  pagination: PaginationInfo;
}

interface ListParams {
  page?: number;
  per_page?: number;
  station_id?: string;
}

const DEFAULT_BASE = "";

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${DEFAULT_BASE}${path}`, {
    headers: { "Content-Type": "application/json" },
    ...init,
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body.error || `API error: ${res.status} ${res.statusText}`);
  }
  if (res.status === 204) return undefined as T;
  return res.json();
}

export async function listChargers(
  params?: ListParams,
): Promise<PaginatedResponse<AdminChargerDto>> {
  const qs = new URLSearchParams();
  if (params?.page) qs.set("page", String(params.page));
  if (params?.per_page) qs.set("per_page", String(params.per_page));
  if (params?.station_id) qs.set("station_id", params.station_id);
  const q = qs.toString();
  return request<PaginatedResponse<AdminChargerDto>>(
    `/api/v1/chargers${q ? `?${q}` : ""}`,
  );
}

export async function getCharger(id: string): Promise<AdminChargerDto> {
  return request<AdminChargerDto>(`/api/v1/chargers/${encodeURIComponent(id)}`);
}

export async function createCharger(
  data: CreateChargerRequest,
): Promise<AdminChargerDto> {
  return request<AdminChargerDto>("/api/v1/chargers", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function updateCharger(
  id: string,
  data: UpdateChargerRequest,
): Promise<AdminChargerDto> {
  return request<AdminChargerDto>(
    `/api/v1/chargers/${encodeURIComponent(id)}`,
    {
      method: "PUT",
      body: JSON.stringify(data),
    },
  );
}

export async function deleteCharger(id: string): Promise<void> {
  return request<void>(
    `/api/v1/chargers/${encodeURIComponent(id)}`,
    { method: "DELETE" },
  );
}
